File: c/heap.c
#include <stdio.h>
#include <stdlib.h>
#include "heap.h"

void ajout(tas **t, arbreCompression *a) {            
    if (*t == NULL) {
        *t = (tas*)malloc(sizeof(tas));        
        (*t)->fg = NULL;
        (*t)->noeud = a;
        (*t)->fd = NULL;        
    } else {
        if ((*a).frequency < (*t)->noeud->frequency) {
            tas *temp1 = (*t)->fg;
            arbreCompression *temp2 = malloc(sizeof(arbreCompression));
            temp2 = (*t)->noeud;
            (*t)->fg = (*t)->fd;
            (*t)->noeud = a;
            (*t)->fd = temp1;
            ajout(&((*t)->fd), temp2);
        }
        else{
            tas *temp = (*t)->fg;
            (*t)->fg = (*t)->fd;
            (*t)->fd = temp;
            ajout(&((*t)->fd), a);
        }
    }
}

void ajoutListe(tas **t, arbreCompression **liste, unsigned int taille){
    for(unsigned int i=0;i<taille;i++){
        ajout(t,liste[i]);
    }
}

void popRacine(tas **t){
    if((*t)->fg == NULL && (*t)->fd == NULL){
        *t = NULL;
    }
    else if((*t)->fd == NULL){
        *t = (*t)->fg;
    }
    else if((*t)->fg == NULL){
        *t = (*t)->fd;
    }
    else{
        arbreCompression *ag = (*t)->fg->noeud;
        arbreCompression *ad = (*t)->fd->noeud;
        if((*ag).frequency < (*ad).frequency){
            (*t)->noeud = ag;
            popRacine(&((*t)->fg));
        }
        else{
            (*t)->noeud = ad;
            popRacine(&((*t)->fd));
        }
    } 
}

arbreCompression *popMinElement(tas **t){
    arbreCompression *res = (*t)->noeud;
    popRacine(t);
    return res;
}

File: ocaml/heap.ml
(*Module Heap*)
(*Définition des types*)
type 'a couple = Element of 'a * int;;
type 'a tas = Vide | Noeud of 'a tas * 'a couple * 'a tas;;

(*Fonctions usuelles d'une structure de tas*)
let creerTas () = Vide;;

let rec ajout t (Element(elt,freq)) = match t with
|Vide -> Noeud(Vide,Element(elt,freq),Vide)
|Noeud(tg,Element(a,b),td) when freq < b -> Noeud(td,Element(elt,freq),ajout tg (Element(a,b)))
|Noeud(tg,Element(a,b),td) -> Noeud(td,Element(a,b),ajout tg (Element(elt,freq)));;

let rec ajoutListe t liste = match liste with
	|[] -> t
	|a::q -> ajoutListe (ajout t a) q;;

let ajoutListe2 liste = 
	let rec aux t l = match l with
	|[] -> t
	|a::q -> aux (ajout t a) q
	in aux Vide liste;;

let rec popRacine t = match t with
|Vide -> failwith"Le tas est vide"
|Noeud(Vide,_,Vide) -> Vide
|Noeud(tg,_,Vide) -> tg
|Noeud(Vide,_,td) -> td
|Noeud(Noeud(tgg,Element(eltg,freqg),tgd),_,Noeud(tdg,Element(eltd,freqd),tdd)) when freqg < freqd -> Noeud(popRacine (Noeud(tgg,Element(eltg,freqg),tgd)),Element(eltg,freqg),Noeud(tdg,Element(eltd,freqd),tdd))
|Noeud(Noeud(tgg,Element(eltg,freqg),tgd),_,Noeud(tdg,Element(eltd,freqd),tdd)) -> Noeud(Noeud(tgg,Element(eltg,freqg),tgd),Element(eltd,freqd),popRacine(Noeud(tdg,Element(eltd,freqd),tdd)));;

let popMinElement t = match t with
|Vide -> failwith "Le tas est vide"
|Noeud(_,a,_) -> (a, popRacine t);;
